### PR TITLE
Adding IdentityMap and ComposedMap classes

### PR DESCRIPTION
**Pull request by michael-c-brennan**
IdentityMap: Implements rectangular maps of the form $I(x_1, x_2) = x_2$

ComposedMap: Given square maps $T_1, \dots, T_n$, implements $T = T_n \circ \dots \circ T_1$. 

Note: Gradient functions for the ComposedMap class require many intermediate evaluations $T_k \circ \dots \circ T_1(x)$ for  $k \le n$. Currently, when these evaluations are needed we recompute them (rather than storing them), which gets slow for large $n$. The plan is to implement checkpointing (i.e., storing $T_k \circ \dots \circ T_1(x)$ for different $k$ to reduce cost of re-computation).

### CODE REVIEW

**Review by mparno — CHANGES_REQUESTED** [review]
Overall this looks good.  I have a few minor requests that will be needed to compile the code with a GPU backend, but everything should be ready to merge when those are addressed.

File: src/IdentityMap.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+template<typename MemorySpace>
+void IdentityMap<MemorySpace>::GradientImpl(StridedMatrix<const double, MemorySpace> const& pts,  
+                            StridedMatrix<const double, MemorySpace> const& sens,
+                            StridedMatrix<double, MemorySpace>              output)
+{
+
+
+    //zero until inputDim-outDim
+    for(unsigned int i=0; i<int(this->inputDim - this->outputDim); ++i)
```
[the remark this answers is not in the document]
Comment (michael-c-brennan) [reply src/IdentityMap.cpp]: Fixed!

File: MParT/ConditionalMapBase.h [hunk 2]
```diff
@@ -28,6 +34,9 @@ namespace mpart {
     class ConditionalMapBase : public ParameterizedFunctionBase<MemorySpace>{
 
         friend class TriangularMap<MemorySpace>;
+        friend class ComposedMap<MemorySpace>;
```
[the remark this answers is not in the document]
Comment (michael-c-brennan) [reply MParT/ConditionalMapBase.h]: Fixed!

File: MParT/TriangularMap.h [hunk 3]
```diff
@@ -51,8 +51,10 @@ class TriangularMap : public ConditionalMapBase<MemorySpace>
     */
     using ConditionalMapBase<MemorySpace>::SetCoeffs;
     virtual void SetCoeffs(Kokkos::View<double*, Kokkos::HostSpace> coeffs) override;
+    virtual void WrapCoeffs(Kokkos::View<double*, Kokkos::HostSpace> coeffs) override;
     #if defined(MPART_ENABLE_GPU)
     virtual void SetCoeffs(Kokkos::View<double*, Kokkos::DefaultExecutionSpace::memory_space> coeffs) override;
+    virtual void WrapCoeffs(Kokkos::View<double*, Kokkos::HostSpace> coeffs) override;
```
[the remark this answers is not in the document]
Comment (michael-c-brennan) [reply MParT/TriangularMap.h]: Fixed!

File: src/IdentityMap.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+
+
+template<typename MemorySpace>
+void IdentityMap<MemorySpace>::LogDeterminantImpl(StridedMatrix<const double, MemorySpace> const& pts,
+                                                    StridedVector<double, MemorySpace>              output)
+{
+
+    // Add to logdet of full map
+    for(unsigned int j=0; j<output.size(); ++j)
```
[the remark this answers is not in the document]
Comment (michael-c-brennan) [reply src/IdentityMap.cpp]: Fixed!

File: src/IdentityMap.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+{
+    assert(false);
+}
+
+template<typename MemorySpace>
+void IdentityMap<MemorySpace>::LogDeterminantInputGradImpl(StridedMatrix<const double, MemorySpace> const& pts, 
+                                            StridedMatrix<double, MemorySpace>              output)
+{   
+    // Add to logdet of full map
+    for(unsigned int i=0; i<output.extent(0); ++i)
```
[the remark this answers is not in the document]
Comment (michael-c-brennan) [reply src/IdentityMap.cpp]: Fixed!

**Comment by michael-c-brennan** [comment]
@mparno I think I fixed everything.

**Review by mparno — APPROVED** [review]

**Pull request merged**